httpcompression middleware: warn of unexpected and prevent manual unupported encodings

cakemd opened the pull request:
This change is made for solving issue [#4697](https://github.com/scrapy/scrapy/issues/4697).

PR [#4697](https://github.com/scrapy/scrapy/pull/5480) is used in terms of fast understanding where the code base must be changed. Please, inform me if this PR must be to that branch first and not directly to master of scrapy/scrapy.

Please, not that in initial version of this PR, the body becomes empty bytes in case of b"br" is used in response but not supported by executing machine. Please, inform me if it's not correct approach.

Will be happy to get review.

wRAR commented on scrapy/downloadermiddlewares/httpcompression.py:
```diff
@@ -17,6 +18,9 @@
 ACCEPTED_ENCODINGS: List[bytes] = [b"gzip", b"deflate"]
+logger = logging.getLogger(__name__)
+
+ACCEPTED_ENCODINGS = [b"gzip", b"deflate"]
```
This looks like a merge error.

cakemd replied:
Yup. I'm working on some more changes. Many tests got broken, fixed that already. Polishing next commit now.

wRAR commented on scrapy/downloadermiddlewares/httpcompression.py:
```diff
@@ -85,6 +91,19 @@ def process_response(
 
+    def _raise_unsupported_compressors(self, request: Request):
+        encodings = request.headers.getlist("Accept-Encoding")
+        unsupported = [key for key in encodings if key not in ACCEPTED_ENCODINGS]
```
You need to split the header value.

cakemd commented on scrapy/downloadermiddlewares/httpcompression.py:
```diff
@@ -99,8 +118,15 @@ def _decode(self, body: bytes, encoding: bytes) -> bytes:
+            if b"br" in ACCEPTED_ENCODINGS:
+                body = brotli.decompress(body)
+            else:
+                body = bytes()
```
Yes, was still thinking what to return. Will change.

wRAR commented:
So, this raises exceptions for requests with unsupported encodings declared in Accept-Encoding. I'm not sure if this is the best course of action. In the previous issues/PRs there were also such options suggested as logging a warning and removing the unsupported encodings from the header.

In any case, the warning/error logged/raised should clearly explain what is the problem in trhe case of brotli and zstd: that the user should install a module to get it supported.

cakemd commented:
@wRAR we can change raising an error to logging error message, not a problem. I just followed @Gallaecio 's issue description.

I've changed PR already, so it fires exception only if it's b"br" manually added by user and not supported by machine.
(Initially, I added raising error for any not supported encoding, but I faced failed tests caused by some encodings that are not present in ACCEPTED _ENCODINGS but are in use).

wRAR commented on scrapy/downloadermiddlewares/httpcompression.py:
```diff
@@ -85,6 +93,22 @@ def process_response(
+
+    def _raise_unsupported_compressors(self, request: Request):
+        encodings = request.headers.getlist("Accept-Encoding")
+        if encodings and len(encodings):
```
I think the `len` check here is not needed.

wRAR commented on scrapy/downloadermiddlewares/httpcompression.py:
```diff
@@ -85,6 +93,22 @@ def process_response(
+    def _raise_unsupported_compressors(self, request: Request):
+        encodings = request.headers.getlist("Accept-Encoding")
+        if encodings and len(encodings):
+            encodings = encodings.pop().split(ENCODINGS_DELIMETER)
```
Not sure why is this doing `encodings.pop()` as you could use `encodings[1]`, but also this should iterate over `encodings` instead of taking the first item.
